File: gen_code.rb
require 'yaml'
require 'pathname'
require 'optparse'

STACK       = "stack"
ENVIRONMENT = "env"
CODE        = "code"
DUMP        = "dump"
VMEXEC      = "exec"
FUNCPREFIX  = "get_"
RET         = "ret"
TRANSPRIM = "transform_primitive"

DESTRUCTURING_RULES = {
  "int" => lambda {|v| "IntegerConstant(#{v})"},
  "bool" => lambda {|v| "BooleanConstant(#{v})"},
  "context" => lambda {|v| "Context(#{v})"},
  "float" => lambda {|v| "FloatConstant(#{v})"},
  "horz" => lambda {|v| "Horz(#{v})"},
  "vert" => lambda {|v| "Vert(#{v})"},
  "length" => lambda {|v| "LengthConstant(#{v})"},
  "math" => lambda {|v| "MathValue(#{v})"},
  "path_value" => lambda {|v| "PathValue(#{v})"},
  "prepath" => lambda {|v| "PrePathValue(#{v})"},
  "regexp" => lambda {|v| "RegExpConstant(#{v})"},
}

def default_false b
  if b == nil then false else b end
end

def gen_prims tag
  YAML.load_stream(ARGF.read) do |inst|
    if default_false(inst[tag]) && inst["name"] != nil then
      len = inst["params"].length
      args = []
      for i in 1..len
        args.push "_v#{i}"
      end

      puts "        (\"#{inst["name"]}\","

      puts "          begin"
      inst["type"].each_line do |line|
        puts "            #{line}"
      end
      puts "          end,"

      puts "          lambda#{len} (fun #{args.join ' '} -> #{inst["inst"]}(#{args.join ', '}))"
      puts "        );"
    end
  end
end

def gen_pdf_mode_prims
  gen_prims("is-pdf-mode-primitive")
end

def gen_text_mode_prims
  gen_prims("is-text-mode-primitive")
end

def gen_interps_0
  YAML.load_stream(ARGF.read) do |inst|
    if (inst["is-pdf-mode-primitive"] || inst["is-text-mode-primitive"]) && !default_false(inst["no-interp"]) then
      tmpn = 0
      astargs = []
      valueidents = []
      otheridents = []

      inst["params"].each do |param|
        if param.is_a?(String) then
          astargs.push "_ast#{tmpn}"
          valueidents.push [param, "_ast#{tmpn}"]
          tmpn = tmpn + 1
        elsif param.is_a?(Hash) then
          ident = param.keys[0]
          type = param.values[0]
          astargs.push "_ast#{tmpn}"
          otheridents.push [ident, type, "_ast#{tmpn}"]
          tmpn = tmpn + 1
        end
      end if inst["params"]

      puts "  | #{inst["inst"]}(#{astargs.join ', '}) ->"
      valueidents.each do |pair|
        ident, astident = pair
        puts "      let #{ident} = interpret_0 env #{astident} in"
      end
      otheridents.each do |tri|
        ident, type, astident = tri
        puts "      let #{ident} = #{FUNCPREFIX}#{type} (interpret_0 env #{astident}) in"
      end
      puts "      let reducef = reduce_beta_list in" if inst["needs-reducef"]
      puts "        begin"
      if inst["separated"] then
        inst["code-interp"].each_line do |line|
          puts "          #{line}"
        end
      else
        inst["code"].each_line do |line|
          puts "          #{line}"
        end
      end
      puts "        end"
      puts ""
    end
  end
end

def gen_vminstrs
  YAML.load_stream(ARGF.read) do |inst|
    tmpn = 0
    destruct = []
    funcapp = []

    inst["params"].each do |param|
      if param.is_a?(String) then
        destruct.push param
      elsif param.is_a?(Hash) then
        pat = param.keys[0]
        type = param.values[0]
        rule = DESTRUCTURING_RULES[type]
        if rule then
          destruct.push (rule.call pat)
        else
          destruct.push "_tmp#{tmpn}"
          funcapp.push [pat, type, "_tmp#{tmpn}"]
          tmpn = tmpn + 1
        end
      end
    end if inst["params"]

    if inst["fields"] != nil then
      fieldnames = inst["fields"].collect{|e| e.keys[0]}
      puts "  | Op#{inst["inst"]}(#{fieldnames.join ', '}) ->"
    else
      puts "  | Op#{inst["inst"]} ->"
    end
    puts "      begin"
    puts "        match #{STACK} with" if inst["params"]

    puts "        | #{destruct.reverse.join(" :: ")} :: #{STACK} ->" if destruct.length > 0

    funcapp.each do |app|
      dest, func, src = app
      puts "            let #{dest} = #{FUNCPREFIX}#{func} #{src} in"
    end
    if inst["needs-reducef"] then
      puts "            let reducef = exec_application #{ENVIRONMENT} in"
    end
    if (inst["is-pdf-mode-primitive"] || inst["is-text-mode-primitive"]) then
      puts "            let #{RET} ="
    else
      puts "            begin"
    end
    inst["code"].each_line do |line|
      puts "              #{line}"
    end
    if (inst["is-pdf-mode-primitive"] || inst["is-text-mode-primitive"]) then
      puts "            in #{VMEXEC} (#{RET} :: #{STACK}) #{ENVIRONMENT} #{CODE} #{DUMP}"
    else
      puts "            end"
    end
    puts ""
    puts "        | _ -> report_bug_vm \"invalid argument for Op#{inst["inst"]}\"" if inst["params"]
    puts "      end"
    puts ""
  end
end

def gen_insttype
  puts "and instruction ="
  YAML.load_stream(ARGF.read) do |inst|
    if inst["fields"] != nil then
      fieldtypes = inst["fields"].collect{|e| e.values[0]}
      puts "  | Op#{inst["inst"]} of #{fieldtypes.join ' * '}"
    else
      puts "  | Op#{inst["inst"]}"
    end

    puts "      [@printer (fun fmt _ -> Format.fprintf fmt \"Op#{inst["inst"]}(...)\")]" if inst["suppress-pp"]
    puts "      [@printer (#{inst["custom-pp"]})]" if inst["custom-pp"]
  end
  puts "  [@@deriving show { with_path = false; }]"
end

def gen_attype
  YAML.load_stream(ARGF.read) do |inst|
    if (inst["is-pdf-mode-primitive"] || inst["is-text-mode-primitive"]) && !default_false(inst["no-ircode"]) then
      if inst["params"] != nil then
        puts "  | #{inst["inst"]} of #{(["abstract_tree"] * inst["params"].length).join ' * '}"
      else
        puts "  | #{inst["inst"]}"
      end
    end
  end
end

def gen_ircases
  YAML.load_stream(ARGF.read) do |inst|
    if (inst["is-pdf-mode-primitive"] || inst["is-text-mode-primitive"]) && !default_false(inst["no-ircode"]) then
      params = [*1..inst["params"].length].collect{|n| "p"+n.to_s}

      puts "    | #{inst["inst"]}(#{params.join ', '}) ->"
      puts "        #{TRANSPRIM} env [#{params.join '; '}] Op#{inst["inst"]}"
      puts ""
    end
  end
end

def f(inst, key, label)
   v = inst[key]
   unless v.nil? then
      vv = if block_given? then yield v else v end
      puts "        ~#{label}:#{vv}"
   end
end

def gen_ml
   print <<EOF
(**
   {1 SATySFi virtual machine instruction definitions}

   This file is a part of [gencode.exe].
   [gencode.exe] generates OCaml source code and
   type declarations included from

   - [types_.cppo.ml],
   - [ir_.cppo.ml],
   - [evaluator_.cppo.ml], and
   - [vm_.cppo.ml].

   To add a new primitive,

   + Add a new instruction definition to this file.
      ([is_pdf_mode_primitive] or [is_text_mode_primitive] should be [true].)
   + Add a new entry to [primitives.ml].
*)

EOF
   puts 'let def ='
   puts '  Instruction.('
   n = 0
   YAML.load_stream(ARGF.read) do |inst|
      sep = (n == 0) ? '[' : ';'
      fs = inst['fields']&.collect{|h|
         k = h.keys[0]
         v = h.values[0]
         "          field #{k.dump} ~type_:#{v.dump};\n"
      }&.join('') || ''
      ps = inst['params']&.collect{|v|
         case v
         when String then
            "          param #{v.dump};\n"
         when Hash then
            k = v.keys[0]
            t = v.values[0]
            "          param #{k.dump} ~type_:#{t.dump};\n"
         end
      }&.join('') || ''
      puts "    #{sep} inst #{inst['inst'].dump}"
      f(inst, 'name', 'name') {|v| v.dump }
      if inst['type'] then
         puts "        ~type_:{|"
         puts inst['type']
         puts "|}"
      end
      puts "        ~fields:[\n#{fs}        ]"
      puts "        ~params:[\n#{ps}        ]"
      f(inst, 'is-pdf-mode-primitive', 'is_pdf_mode_primitive')
      f(inst, 'is-text-mode-primitive', 'is_text_mode_primitive')
      f(inst, 'needs-reducef', 'needs_reducef')
      if inst['suppress-pp'] then
         puts "        ~pp:Simple"
      elsif inst['custom-pp'] then
         v = inst['custom-pp']
         puts "        ~pp:(Custom #{v.dump})"
      end
      f(inst, 'no-ircode', 'no_ircode')
      f(inst, 'no-interp', 'no_interp')
      if inst['code-interp'] then
         puts "        ~code_interp:{|"
         puts inst['code-interp']
         puts "|}"
      end
      puts "        ~code:{|"
      puts inst['code']
      puts "|}"
      n += 1
   end
   puts '    ])'
end

opt = OptionParser.new

func = nil

opt.on('--gen-vm') {|v| func = method(:gen_vminstrs) }
opt.on('--gen-ir') {|v| func = method(:gen_ircases) }
opt.on('--gen-insttype') {|v| func = method(:gen_insttype) }
opt.on('--gen-attype') {|v| func = method(:gen_attype) }
opt.on('--gen-interps-0') {|v| func = method(:gen_interps_0) }
opt.on('--gen-pdf-mode-prims') {|v| func = method(:gen_pdf_mode_prims) }
opt.on('--gen-text-mode-prims') {|v| func = method(:gen_text_mode_prims) }
opt.on('--ml') {|v| func = method(:gen_ml) }

opt.parse!(ARGV)

func.call

File: tools/gencode/gencode.ml
open U

(*
let failwith _ = ()  (* TEMPORARY *)
*)

module Const = struct
  let stack = "stack"

  let environment = "env"

  let code = "code"

  let dump = "dump"

  let vmexec = "exec"

  let func_prefix = "get_"

  let ret = "ret"

  let make_entry = "make_entry"

  let trans_prim = "transform_0_primitive"

  let destructuring_rules =
    let open Printf in
    [ "int", sprintf "BaseConstant(BCInt(%s))"
    ; "bool", sprintf "BaseConstant(BCBool(%s))"
    ; "context", sprintf "Context(%s)"
    ; "float", sprintf "BaseConstant(BCFloat(%s))"
    ; "horz", sprintf "BaseConstant(BCHorz(%s))"
    ; "vert", sprintf "BaseConstant(BCVert(%s))"
    ; "length", sprintf "BaseConstant(BCLength(%s))"
    ; "math", sprintf "MathValue(%s)"
    ; "path_value", sprintf "BaseConstant(BCPath(%s))"
    ; "prepath", sprintf "BaseConstant(BCPrePath(%s))"
    ; "regexp", sprintf "BaseConstant(BCRegExp(%s))"
    ]
end


let is_pdf_mode_primitive def =
  def.Instruction.is_pdf_mode_primitive


let is_text_mode_primitive def =
  def.Instruction.is_text_mode_primitive


let gen_prims is_prims =
  let open Instruction in
  Vminst.def |> List.iter (function
  | {
      name = Some(name);
      type_ = Some(t);
      inst;
      params;
      _
    } as def  when is_prims def ->
      let args = params |> List.mapi (fun i _ -> "_v%d" @% i + 1) in
      let len = List.length args in
      puts "        (\"%s\"," name;
      puts "          begin";
      puts "            %s" @@ Type.to_code t;
      puts "          end,";
      puts "          lambda%d (fun %s -> %s(%s))"
        len (String.concat " " args) inst (String.concat ", " args);
      puts "        );"

  | _ ->
      ()
  )


let gen_pdf_mode_prims () =
  gen_prims is_pdf_mode_primitive


let gen_text_mode_prims () =
  gen_prims is_text_mode_primitive


let gen_interps_1 () =
  let open Instruction in
  Vminst.def |> List.iter (function
  | {
      no_interp = false;
      inst;
      params;
      _
    } as def  when is_primitive def ->
      let astargs = params |> List.mapi (fun i _ -> "_ast%d" @% i) in
      let codeargs = params |> List.mapi (fun i _ -> "_code%d" @% i) in
      puts "  | %s(%s) ->" inst (String.concat ", " astargs);
      List.iter2 (fun codearg astarg ->
        puts "      let %s = interpret_1 env %s in" codearg astarg;
      ) codeargs astargs;
      puts "      Cd%s(%s)" inst (String.concat ", " codeargs);
      puts ""

  | _ ->
      failwith "[gen_interps_1] not a primitive."
  )


let gen_interps_0 () =
  let open Instruction in
  Vminst.def |> List.iter (function
  | {
      no_interp = false;
      inst;
      params;
      needs_reducef;
      code_interp;
      code;
      _
    } as def  when is_primitive def ->
      let astargs = params |> List.mapi (fun i _ -> "_ast%d" @% i) in
      puts "  | %s(%s) ->" inst (String.concat ", " astargs);
      List.combine params astargs |> List.iter (function
      | ({ Param.name; type_ = None }, astident) ->
          puts "      let %s = interpret_0 env %s in"
            name astident

      | _ ->
          ()
      );
      List.combine params astargs |> List.iter (function
      | ({ Param.name; type_ = Some t }, astident) ->
          puts "      let %s = %s%s (interpret_0 env %s) in"
            name Const.func_prefix t astident

      | _ ->
          ()
      );
      if needs_reducef then begin
        puts "      let reducef = reduce_beta_list in"
      end;
      puts "        begin";
      default code code_interp |> split_lines |> List.iter
        (puts "          %s");
      puts "        end";
      puts ""

  | _ ->
      failwith "[gen_interps_0] not a primitive."
  )


let make_entry_pattern name =
  "(" ^ name ^ ", _)"


let gen_vminstrs () =
  let open Instruction in
  Vminst.def |> List.iter (function
  | {
      inst;
      params;
      fields;
      needs_reducef;
      code;
      _
    } as def ->
      let i = ref 0 in
      let ps =
        params |> List.map (function
        | { Param.name; type_ = None } as p ->
            (p, name, false)

        | { Param.name; type_ = Some typ } as p ->
            begin
              match List.assoc_opt typ Const.destructuring_rules with
              | Some(rule) ->
                  (p, rule name, false)

              | None ->
                  let tmp = "_tmp%d" @% !i in
                  i := !i + 1;
                  (p, tmp, true)
            end
        )
      in
      let destruct = ps |> List.rev_map (fun (_, x, _) -> x) in
      let funcapp = ps |> List.filter (fun (_, _, x) -> x) in
      begin
        match fields with
        | [] ->
            puts "  | Op%s ->" inst

        | _ :: _ ->
            failwith "[gen_vminstrs] fields should be empty."
(*
        | fs ->
            puts "  | Op%s(%s) ->" inst @@
              String.concat ", " @@ List.map Field.name fs
*)
      end;
      puts "      begin";
      if not @@ nullp params then begin
        puts "        match %s with" Const.stack
      end;
      begin
        match destruct with
        | [] ->
            ()

        | ds ->
            puts "        | %s :: %s ->" (String.concat " :: " (ds |> List.map make_entry_pattern)) Const.stack
      end;
      funcapp |> List.iter (function
      | ({ Param.name = dest; type_ = Some(func) }, src, _) ->
          puts "            let %s = %s%s %s in"
            dest Const.func_prefix func src

      | _ ->
          ()   (* hmm... *)
      );
      if needs_reducef then begin
        puts "            let reducef = exec_application %s in"
          Const.environment
      end;
      let print_code () =
        split_lines code |> List.iter
          (puts "              %s");
      in
      if is_primitive def then begin
        puts "            let %s =" Const.ret;
        print_code ();
        puts "            in %s (%s %s :: %s) %s %s %s"
          Const.vmexec Const.make_entry Const.ret Const.stack Const.environment Const.code Const.dump
      end else begin
        puts "            begin";
        print_code ();
        puts "            end"
      end;
      puts "";
      if not @@ nullp params then begin
        puts "        | _ -> report_bug_vm \"invalid argument for Op%s\"" inst
      end;
      puts "      end";
      puts ""
  )


let gen_insttype () =
  let open Instruction in
  Vminst.def |> List.iter (function
  | {
      inst;
      fields;
      pp;
      name;
      type_;
      _
    } as def  when is_primitive def ->
      begin
        match fields with
        | [] ->
            puts "  | Op%s" inst;
            begin match name, type_ with
              | Some name, Some t ->
                (* use @ocaml.doc attribute to avoid puzzling over escaping *)
                puts "    [@ocaml.doc %S]"
                  (Printf.sprintf "[%s : %s]" (Name.show name) (Type.to_string t))
              | _, _ -> ()
            end

        | _ :: _ ->
            failwith "[gen_insttype] fields should be empty."
      end;
      begin
        match pp with
        | Default ->
            ()

        | Simple ->
            puts "      [@printer (fun fmt _ -> Format.fprintf fmt \"Op%s(...)\")]"
              inst

        | Custom pp ->
            puts "      [@printer (%s)]" pp
      end

  | def ->
      failwith ("[gen_insttype] not a primitive: " ^ def.inst)
  );
  puts "  [@@deriving show { with_path = false; }]"


let gen_unliftcode () =
  let open Instruction in
  Vminst.def |> List.iter (function
  | {
      no_ircode = false;
      inst;
      params;
      _
    } as def  when is_primitive def ->
      begin
        match params with
        | [] ->
            puts "  | Cd%s -> %s" inst inst

        | ps ->
            let cdargs = List.mapi (fun i _ -> "code" ^ string_of_int i) ps in
            let args = List.map (fun s -> "aux " ^ s) cdargs in
            puts "  | Cd%s(%s) -> %s(%s)"
              inst
              (String.concat ", " cdargs)
              inst
              (String.concat ", " args)
      end

  | _ ->
      failwith "[unlift_code] not a primitive."
  )

let gen_codetype () =
  let open Instruction in
  Vminst.def |> List.iter (function
  | {
      no_ircode = false;
      inst;
      params;
      _
    } as def  when is_primitive def ->
      begin
        match params with
        | [] ->
            puts "  | Cd%s" inst

        | ps ->
            puts "  | Cd%s of %s" inst
              (String.concat " * " (List.map (const "code_value") ps))
      end

  | _ ->
      failwith "[gen_codetype] not a primitive."
  )


let gen_attype () =
  let open Instruction in
  Vminst.def |> List.iter (function
  | {
      no_ircode = false;
      inst;
      params;
      _
    } as def  when is_primitive def ->
      begin
        match params with
        | [] ->
            puts "  | %s" inst

        | ps ->
            puts "  | %s of %s"
              inst
              (String.concat " * "
               @@ List.map (const "abstract_tree") ps)
      end

  | _ ->
      ()
  )


let gen_ircases_0 () =
  let open Instruction in
  Vminst.def |> List.iter (function
  | {
      no_ircode = false;
      inst;
      params;
      _
    } as def  when is_primitive def ->
      let ps = params |> List.mapi (fun i _ -> "p%d" @% i + 1) in
      puts "    | %s(%s) ->"
        inst
        (String.concat ", " ps);
      puts "        %s env [%s] Op%s"
        Const.trans_prim
        (String.concat "; " ps)
        inst;
      puts ""

  | _ ->
      ()
  )


let gen_ircases_1 () =
  let open Instruction in
  Vminst.def |> List.iter (function
  | {
      no_ircode = false;
      inst;
      params;
      _
    } as def  when is_primitive def ->
      let len = List.length params in
      let ps = params |> List.mapi (fun i _ -> "p%d" @% i + 1) in
      let cvs = params |> List.mapi (fun i _ -> "cv%d" @% i + 1) in
      puts "    | %s(%s) ->"
        inst
        (String.concat ", " ps);
      puts "        code%d env (fun %s -> Cd%s(%s)) %s"
        len
        (String.concat " " cvs)
        inst
        (String.concat ", " cvs)
        (String.concat " " ps);
      puts ""

  | _ ->
      ()
  )


let () =
  let opts =
    [
      ("--gen-vm"             , gen_vminstrs       );
      ("--gen-ir-0"           , gen_ircases_0      );
      ("--gen-ir-1"           , gen_ircases_1      );
      ("--gen-insttype"       , gen_insttype       );
      ("--gen-attype"         , gen_attype         );
      ("--gen-codetype"       , gen_codetype       );
      ("--gen-unliftcode"     , gen_unliftcode     );
      ("--gen-interps-0"      , gen_interps_0      );
      ("--gen-interps-1"      , gen_interps_1      );
      ("--gen-pdf-mode-prims" , gen_pdf_mode_prims );
      ("--gen-text-mode-prims", gen_text_mode_prims);
    ]
  in
  let opt = Sys.argv.(1) in
  match List.assoc_opt opt opts with
  | Some(func) -> func ()
  | None       -> failwith @@ Printf.sprintf "unknown option: %s" opt
